### DISCUSSION TITLE
scanned pdf

### DISCUSSION THREAD

**Discussion by 88arvin** [pymupdf/PyMuPDF-Utilities#95] [General]
How can the tabular data in the scanned PDF be converted to a csv?

**Comment by JorjMcKie** [comment]
You must of course OCR the page first. After that, you can extract text (almost) like normal.

You still need **_external_** information to find the position (wrapping rectangle) of the table. This includes but is certainly not limited to things like:
* you may know, that the table starts below some specific text string, and ends before some other specific text string
* you may kow, that no text exists to the left and the right sides of the table
* ...


Be aware, that any drawings like lines around single table cells will not be recognizes by (most of) the OCR program - they don't exist in the OCR output.
So you must find a way to identify column borders yourself. This is not trivial, especially if cell content is not aligned in a particular way.
All this becomes much easier if you can provide "external" knowledge to the script you must write; something like _"column 3 is text left-aligned, column 4 is numeric righ-alinged"_ and so on.

I hope my reasoning has become clear.

Once you have identified column and row borders, things are easy: both informations are lists of float values, which can be used to make rectangles, each representing a table cell. You can extract the text inside a rectangle by using text extraction with the "clip" parameter.

**Reply by 88arvin** [reply]
Thank you for your response.

I have used PP-OCR, and it is performing very well, with an accuracy of approximately 95%. Now the problem is that when I am using the code for PDF, it is giving the following error: AttributeError: 'Document' object has no attribute 'pageCount'

**Reply by JorjMcKie** [reply]
Please have a look at the documentation's chapter [Deprecated Names](https://pymupdf.readthedocs.io/en/latest/znames.html). In your example, the property is named `page_count` since a long time.

**Reply by 88arvin** [reply]
Still getting the same error after changing pageCount to page_count.

```
from paddleocr import PaddleOCR, draw_ocr
ocr = PaddleOCR(use_angle_cls=True, lang="en, page_num=0)
result = ocr.ocr(img_path, cls=True)
for idx in range(len(result)):
    res = result[idx]
    for line in res:
        print(line)

import fitz
from PIL import Image
import cv2
import numpy as np
imgs = []
with fitz.open(img_path) as pdf:
    for pg in range(0, pdf.page_count):
        page = pdf[pg]
        mat = fitz.Matrix(2, 2)
        pm = page.getPixmap(matrix=mat, alpha=False)
        # if width or height > 2000 pixels, don't enlarge the image
        if pm.width > 2000 or pm.height > 2000:
            pm = page.getPixmap(matrix=fitz.Matrix(1, 1), alpha=False)

        img = Image.frombytes("RGB", [pm.width, pm.height], pm.samples)
        img = cv2.cvtColor(np.array(img), cv2.COLOR_RGB2BGR)
        imgs.append(img)
for idx in range(len(result)):
    res = result[idx]
    image = imgs[idx]
    boxes = [line[0] for line in res]
    txts = [line[1][0] for line in res]
    scores = [line[1][1] for line in res]
    im_show = draw_ocr(image, boxes, txts, scores, font_path='doc/fonts/simfang.ttf')
    im_show = Image.fromarray(im_show)
    im_show.save('result_page_{}.jpg'.format(idx))```

**Reply by JorjMcKie** [reply]
No, cannot be ... for `page_count`.
But should now be getting the equivalent error for `getPixmap`. This has been renamed too: `getPixmap`->`get_pixmap` - as explained in the documentation chapter referenced above 🙁.

As an aside: you can simplify your iteration somewhat:
```python
for page in pdf:
```
instead of
```python
for pg in range(0, pdf.page_count):
    page = pdf[pg]
```

**Reply by chanpreet90** [reply]
Still getting the same error :(

**Reply by JorjMcKie** [reply]
@chanpreet90 ket me have the exception output please

**Reply by chanpreet90** [reply]
![input](https://user-images.githubusercontent.com/86623814/216767296-97c6896e-6047-4cef-b1f6-268b0519fa79.PNG)

**Reply by JorjMcKie** [reply]
you did not show the complete output (right hand side missing), but apparently package `paddleocr` uses PyMuPDF and contains the deleted attribute name.

**Confirmed:**
This package uses PyMuPDF and contains a wrong dependency info in its requirements text file: PyMuPDF<1.21.0. It should be <1.20.0.

So this is my recommendation:
1. Submit a bug issue to the package
2. **_Either_** downlevel your PyMuPDF to the latest 1.19.* - **_or_** go into you PyMuPDF installation folder and change `__init__.py` and replace lines 481/482 by just one line `restore_aliases()`.
![grafik](https://user-images.githubusercontent.com/8290722/216768083-21ad1971-4b4f-47b6-b856-482695493681.png)

This will cause PyMuPDF to accept the old names again.

**Reply by 88arvin** [reply]
Now, it's functional. Much appreciation
